src/collapse/vsprof.rs

```diff
@@ -0,0 +1,222 @@
+use std::{
+    cmp::Ordering,
+    io::{self, BufRead},
+};
+
+use log::warn;
+
+use crate::collapse::{common::Occurrences, Collapse};
+
+static START_LINE: &str = "Level,Function Name,Number of Calls,Elapsed Inclusive Time %,Elapsed Exclusive Time %,Avg Elapsed Inclusive Time,Avg Elapsed Exclusive Time,Module Name,";
+
+/// A stack collapser for the output of the Visual Studio built in profiler.
+#[derive(Default)]
+pub struct Folder {
+    /// Function entries on the stack in this entry thus far.
+    stack: Vec<(String, usize)>,
+}
+
+impl Collapse for Folder {
+    fn collapse<R, W>(&mut self, mut reader: R, writer: W) -> io::Result<()>
+    where
+        R: std::io::BufRead,
+        W: std::io::Write,
+    {
+        // Skip the header
+        let mut line = Vec::new();
+        if reader.read_until(b'\n', &mut line)? == 0 {
+            warn!("File ended before start of call graph");
+            return Ok(());
+        };
+
+        let header = String::from_utf8_lossy(&line).to_string();
+        if !line_matches_start_line(header.clone()) {
+            return invalid_data_error!("Incorrect header:\n{}", header);
+        }
+
+        // Process the data
+        let mut occurences = Occurrences::new(1);
+        loop {
+            line.clear();
+            if reader.read_until(b'\n', &mut line)? == 0 {
+                break;
+            }
+            let l = String::from_utf8_lossy(&line);
+            let line = l.trim_end();
+            if line.is_empty() {
+                continue;
+            } else {
+                self.on_line(line, &mut occurences)?;
+            }
+        }
+
+        self.write_stack(&mut occurences);
+
+        // Write the results
+        occurences.write_and_clear(writer)?;
+
+        // Reset the state
+        self.stack.clear();
+        Ok(())
+    }
+
+    fn is_applicable(&mut self, input: &str) -> Option<bool> {
+        let mut input = input.as_bytes();
+        let mut line = String::new();
+        if let Ok(n) = input.read_line(&mut line) {
+            if n == 0 {
+                return Some(false);
+            }
+        } else {
+            return Some(false);
+        }
+
+        Some(line_matches_start_line(line))
+    }
+}
+
+impl Folder {
+    // Parse lines like:
+    // Level,Function Name,Number of Calls,...
+    // 6,"System.String.IsNullOrEmpty(string)",4,0.00,0.00,0.00,0.00,"mscorlib.dll",
+    fn on_line(&mut self, line: &str, occurences: &mut Occurrences) -> io::Result<()> {
+        let (depth, remainder) = get_next_number(line)?;
+
+        // Function names are always wrapped in spaces. By trimming the leading double quote, we
+        // know that the next double quote is the double quote closing the function name. Splitting
```

Comment:
Heh, that's unfortunately only true _if_ the generator doesn't ever generate `\"` with the expectation that it _doesn't_ end the string. But I'm guessing it's exceedingly rare that a function name has a `"` in it.

Reply:
In C# a function name cannot have a `"` inside it, and I doubt that there are many (if any) languages which do allow them.

According to the [docs], the Visual Studio profiler only supports `.NET`, `C/C++` and `Javascript`. None of these languages support `'"`s in identifiers, as far as I know, so I feel like it is safe to assume that any `"` encountered is not part of the function name itself.

[docs]: https://docs.microsoft.com/en-us/windows/apps/performance/choose-between-tools